schematic/configuration/configuration.py

```diff
@@ -0,0 +1,247 @@
+            fields (list[str]): The fields in the config
+            allowed_fields (list[str]): The allowed fields in the config
+        """
+        self.message = message
+        self.fields = fields
+        self.allowed_fields = allowed_fields
+        super().__init__(self.message)
+
+    def __str__(self) -> str:
+        """String representation"""
+        return (
+            f"{self.message}; "
+            f"config contains fields: {self.fields}; "
+            f"allowed fields: {self.allowed_fields}"
+        )
+
+
+class Configuration:
+    """
+    This class is used as a singleton by the rest of the package.
+    It is instantiated only once at the bottom of this file, and that
+     instance is imported by other modules
+    """
+
+    def __init__(self) -> None:
+        self.config_path: Optional[str] = None
+        self._parent_directory = os.getcwd()
+        self._synapse_config = SynapseConfig()
+        self._manifest_config = ManifestConfig()
+        self._model_config = ModelConfig()
+        self._google_sheets_config = GoogleSheetsConfig()
+
+    def load_config(self, config_path: str) -> None:
+        """Loads a user created config file and overwrites any defaults  listed in the file
+
+        Args:
+            config_path (str): The path to the config file
+
+        Raises:
+            ConfigNonAllowedFieldError: If there are non allowed fields in the config file
+        """
+        allowed_config_fields = {"asset_store", "manifest", "model", "google_sheets"}
+        config_path = os.path.expanduser(config_path)
+        config_path = os.path.abspath(config_path)
+        self.config_path = config_path
+
+        self._parent_directory = os.path.dirname(config_path)
+
+        with open(config_path, "r", encoding="utf-8") as file:
+            config: dict[str, Any] = yaml.safe_load(file)
+        if not set(config.keys()).issubset(allowed_config_fields):
+            raise ConfigNonAllowedFieldError(
+                "Non allowed fields in top level of configuration file.",
+                list(config.keys()),
+                list(allowed_config_fields),
+            )
+
+        self._manifest_config = ManifestConfig(**config.get("manifest", {}))
+        self._model_config = ModelConfig(**config.get("model", {}))
+        self._google_sheets_config = GoogleSheetsConfig(
+            **config.get("google_sheets", {})
+        )
+        self._set_asset_store(config.get("asset_store", {}))
+
+    def _set_asset_store(self, config: dict[str, Any]) -> None:
+        allowed_config_fields = {"synapse"}
+        if not config:
+            pass
+        if not set(config.keys()).issubset(allowed_config_fields):
+            raise ConfigNonAllowedFieldError(
+                "Non allowed fields in asset_store of configuration file.",
+                list(config.keys()),
+                list(allowed_config_fields),
+            )
+        self._synapse_config = SynapseConfig(**config["synapse"])
```

Comment:
I am not really familiar with the syntax ** here. Could you briefly explain what is the `**` syntax doing here?

Reply:
The ** unpacks the dictionary.

See this [link](https://stackabuse.com/unpacking-in-python-beyond-parallel-assignment/) and scroll down to "https://stackabuse.com/unpacking-in-python-beyond-parallel-assignment/"